portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -28,13 +35,61 @@
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    ArrayList<String> messages = new ArrayList<String>();
-    messages.add("first message");
-    messages.add("second message");
-    messages.add("third message");
-    Gson gson = new Gson();
-    String jsonMessage = gson.toJson(messages);
+    ArrayList<Message> messages = new ArrayList<Message>();
+
+    Query query = new Query("Message");
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    for (Entity entity : results.asIterable()) {
+      String email = (String) entity.getProperty("email");
+      String name = (String) entity.getProperty("name");
+      String message = (String) entity.getProperty("message");
+
+      Message messageObj = new Message(name, email, message);
+      messages.add(messageObj);
+    }
+
+    String jsonMessage = convertToJson(messages);
     response.setContentType("application/json;");
     response.getWriter().println(jsonMessage);
   }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String email = getParameter(request, "email");
+    String name = getParameter(request, "name");
+    String message = getParameter(request, "message");
+
+    if (email.isEmpty() || name.isEmpty() || message.isEmpty()) {
+      response.setContentType("text/html;");
+      response.getWriter().println("<meta http-equiv='refresh' content='3;URL=index.html'>");
+      response.getWriter().println("<p style='color:red;'>All form fields must not be empty</p>");
+      return;
```

Comment:
It's better to do this check in the frontend - just mark them as required form fields. It saves a round trip to the server.

Reply:
I do implement on both side, I believe frontend only implementation is easy to bypass

Reply:
yeah sorry I missed the part in the frontend